File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
const combinationSum = (candidates, target) => {
  candidates.sort((a, b) => a - b);
  const res = [];
  const choose = (pos, sum, arr) => {
    sum += candidates[pos];
    arr = arr.slice();
    arr.push(candidates[pos]);

    if (sum === target) {
      res.push(arr);
      return;
    }

    while (candidates[pos] <= target - sum) {
      choose(pos++, sum, arr)
    }
  }
  for(const pos in candidates) {
    choose(pos, 0, []);
  }
  return res;
};

File: 0039-combination-sum/0039-combination-sum.go
func combinationSum(candidates []int, target int) [][]int {
  slices.Sort(candidates)
  ans := [][]int{}
  var choose func(pos int, sum int, arr []int)
  choose = func(pos int, sum int, arr []int) {
    sum += candidates[pos]
    arr = append(arr, candidates[pos])

    if sum == target {
      ans = append(ans, arr)
    } else {
      for pos < len(candidates) && candidates[pos] <= target - sum {
        choose(pos, sum, slices.Clone(arr))
        pos++
      }
    }
  }
  for pos := range len(candidates) {
    choose(pos, 0, []int{})
  }
  return ans
}